bench_test.go

```diff
@@ -0,0 +1,179 @@
+package dhstore_test
+
+import (
+	"bytes"
+	"context"
+	crand "crypto/rand"
+	"encoding/json"
+	"io"
+	"log"
+	"math"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/ipni/dhstore"
+	"github.com/ipni/dhstore/metrics"
+	"github.com/multiformats/go-multihash"
+	"github.com/stretchr/testify/require"
+)
+
+func BenchmarkDHStore_PutMultihashes_0(b *testing.B) {
+	benchmarkPut(b, 300, 1000, 200)
+}
+
+func BenchmarkDHStore_GetMultihashes_0(b *testing.B) {
+	benchmarkGet(b, 500, 1000, 200)
+}
+
+func benchmarkPut(b *testing.B, chunks, chunkSize, vkLen int) {
+	b.SetBytes(int64(34 + vkLen))
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	dir := b.TempDir()
+
+	store, metr, server := createServer(b, dir)
+	defer store.Close()
+	defer metr.Shutdown(context.Background())
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mhs := randomMultihashes(b, chunks*chunkSize)
+			putMultihashes(b, mhs, chunks, vkLen, server, true)
+		}
+	})
+	b.StopTimer()
+}
+
+func benchmarkGet(b *testing.B, chunks, chunkSize, vkLen int) {
+	b.ResetTimer()
+
+	dir := b.TempDir()
+
+	store, metr, server := createServer(b, dir)
+	defer store.Close()
+	defer metr.Shutdown(context.Background())
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mhs := randomMultihashes(b, chunks*chunkSize)
+			putMultihashes(b, mhs, chunks, vkLen, server, false)
+			rand.Shuffle(len(mhs), func(i, j int) {
+				mhs[i], mhs[j] = mhs[j], mhs[i]
+			})
+			getMultihashes(b, mhs, server, true)
+		}
+	})
+	b.StopTimer()
+}
+
+func putMultihashes(b *testing.B, mhs []multihash.Multihash, chunks, vkLen int, server *http.ServeMux, print bool) []multihash.Multihash {
```

Comment:
These benchmarks benchmark HTTP mux as well as the datastore itself. 

I recommend separating the two. store benchmark is probably the one we are interested in more?

Reply:
In this benchmark I'm interested of measuring what is exposed to the consumer, i.e. the store itself + server.

Reply:
* in this case I recommend [renaming the benchmarks ](https://github.com/ipni/dhstore/pull/32/files#diff-d85e4bbdd3120d2de9d967247dba8011f83f3bcd25c39a30c1ab7a1637be6b14R21)to indicate that it's benchmarking HTTP end to end benchmarking
* My 2p: realistically you need isolated benchmarks for both. Doing end to end is all good and well but that muddies the hotspts in code. If something is slow for an engineer to debug it they would need to benchmark each component anyway. For HTTP part, one would want to mute the storage noise by mocking it such that it returns as fast as possible.